### ISSUE TITLE
Minor timeout bug in Reactor

### ISSUE THREAD

**Issue by wjordan** [puma/puma#2282]
There is a minor timeout bug in `Reactor` affecting unit-test stability, real-world impact is probably minimal.

**To Reproduce**

- Reduce `first_data_timeout` from `2` to `0.01` in `test_timeout_in_data_phase` (just to reproduce the issue quicker- the bug can still affect the existing test): https://github.com/puma/puma/blob/2ded6853cbafa36fdf49b46edd6e1401cb70ef3d/test/test_puma_server.rb#L377
- Run the test repeatedly in a while loop and it will eventually hang:
  ```bash
  while :; do ruby -Ilib:test test/test_puma_server.rb --name /test_timeout_in_data_phase/; done
  ```

**Analysis**

In `Reactor`, the monitor of the `Client` object (`c`) is removed from `@timeouts` when the object becomes readable (has new data available to read): https://github.com/puma/puma/blob/19b2a213f51921a3400854ed500ba9d336f81b7c/lib/puma/reactor.rb#L214-L221

However, the object is only actually moved from the reactor to the thread-pool when `c.try_to_finish` is true (when the header / body is fully buffered): https://github.com/puma/puma/blob/19b2a213f51921a3400854ed500ba9d336f81b7c/lib/puma/reactor.rb#L224-L227

This leaves the edge-case where an object becomes readable but `try_to_finish` returns `false`, leaving the object in the reactor in a state where it will never time out.

**Comment by wjordan** [comment]
Fixed by #2279.

**Comment by dentarg** [comment]
Maybe this is what's happening for https://stackoverflow.com/questions/65290236/heroku-h18-error-in-a-rails-app-on-an-unhandled-route
